Framework/CurveFitting/src/Functions/ProcessBackground.cpp

```diff
@@ -709,49 +702,43 @@ ProcessBackground::filterForBackground(BackgroundFunction_sptr bkgdfunction) {
       WorkspaceFactory::Instance().create("Workspace2D", 4, sizex, sizey));
   for (size_t i = 0; i < sizex; ++i) {
     for (size_t j = 0; j < 4; ++j) {
-      visualws->dataX(j)[i] = domain[i];
+      visualws->mutableX(j)[i] = domain[i];
     }
   }
   for (size_t i = 0; i < sizey; ++i) {
-    visualws->dataY(0)[i] = values[i];
-    visualws->dataY(1)[i] = m_dataWS->readY(m_wsIndex)[i] - values[i];
-    visualws->dataY(2)[i] = posnoisetolerance;
-    visualws->dataY(3)[i] = -negnoisetolerance;
+    visualws->mutableY(0)[i] = values[i];
+    visualws->mutableY(1)[i] = m_dataWS->y(m_wsIndex)[i] - values[i];
+    visualws->mutableY(2)[i] = posnoisetolerance;
+    visualws->mutableY(3)[i] = -negnoisetolerance;
   }
   setProperty("UserBackgroundWorkspace", visualws);
 
   // Filter for background
-  std::vector<double> vecx, vecy, vece;
+  std::vector<size_t> selectedIndexes;
   for (size_t i = 0; i < domain.size(); ++i) {
-    // double y = m_dataWS->readY(m_wsIndex)[i];
-    // double theoryy = values[i]; y-theoryy
-    double purey = visualws->readY(1)[i];
+    double purey = visualws->y(1)[i];
     if (purey < posnoisetolerance && purey > -negnoisetolerance) {
-      // Selected
-      double x = domain[i];
-      double y = m_dataWS->readY(m_wsIndex)[i];
-      double e = m_dataWS->readE(m_wsIndex)[i];
-      vecx.push_back(x);
-      vecy.push_back(y);
-      vece.push_back(e);
+      selectedIndexes.push_back(i);
     }
   }
-  g_log.information() << "Found " << vecx.size() << " background points out of "
-                      << m_dataWS->readX(m_wsIndex).size()
-                      << " total data points. "
+  size_t wsSize = selectedIndexes.size();
+  g_log.information() << "Found " << wsSize << " background points out of "
+                      << m_dataWS->x(m_wsIndex).size() << " total data points. "
                       << "\n";
 
   // Build new workspace for OutputWorkspace
   size_t nspec = 3;
   Workspace2D_sptr outws =
       boost::dynamic_pointer_cast<DataObjects::Workspace2D>(
-          API::WorkspaceFactory::Instance().create("Workspace2D", nspec,
-                                                   vecx.size(), vecy.size()));
-  for (size_t i = 0; i < vecx.size(); ++i) {
-    for (size_t j = 0; j < nspec; ++j)
-      outws->dataX(j)[i] = vecx[i];
-    outws->dataY(0)[i] = vecy[i];
-    outws->dataE(0)[i] = vece[i];
+          API::WorkspaceFactory::Instance().create("Workspace2D", nspec, wsSize,
+                                                   wsSize));
+  for (size_t i = 0; i < wsSize; ++i) {
+    size_t index = selectedIndexes[i];
+    for (size_t j = 0; j < nspec; ++j) {
+      outws->mutableX(j)[i] = domain[index];
+      outws->mutableY(0)[i] = m_dataWS->y(m_wsIndex)[index];
+      outws->mutableE(0)[i] = m_dataWS->e(m_wsIndex)[index];
```

Comment:
I based it off the old code and since I did not know what the intention of doing that was previously, I left it in. I feel that the old code may be incorrect however since the test that covers this function never actually checks the contents `outws`, so the test would pass regardless of what was being passed to the workspace.